### PR TITLE
Validate fields name

### PR DESCRIPTION
**Pull request by sabaghadar**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/35579

## Description
validate incident field name prefix is pack name

## Screenshots
Paste here any images that will help the reviewer

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: demisto_sdk/commands/common/errors.py [hunk 1]
```diff
@@ -1661,3 +1662,16 @@ def xsoar_config_file_is_not_json(file_path):
...
+    @staticmethod
+    @error_code_decorator
+    def invalid_incident_field_prefix(field_name):
+        return f"Field name: {field_name} is invalid. Field name must start with the relevant pack name."
+
+
+    @staticmethod
+    def suggest_fix_field_name(field_name, pack_name):
+        return f"To fix the problem, add pack name prefix to the field name." \
+               f"you can use the pack name or one of the prefixes in packe.metadat itemPrefix. Example: {pack_name} {field_name}." \
```
Comment (roysagi) [inline demisto_sdk/commands/common/errors.py:1675]:
```suggestion
               f"you can use the pack name or one of the prefixes found in the itemPrefix field in the pack_metadata. Example: {pack_name} {field_name}." \
```

File: demisto_sdk/commands/common/errors.py [hunk 2]
```diff
@@ -1661,3 +1662,16 @@ def xsoar_config_file_is_not_json(file_path):
...
+               "Also make sure to update the field id and cliName accordingly." \
+               f"or add the pack name to the ignored packs"
```
Comment (roysagi) [inline demisto_sdk/commands/common/errors.py:1677]: this option should not be exposed to the user.

File: demisto_sdk/commands/common/hook_validations/incident_field.py [hunk 3]
```diff
@@ -397,3 +400,24 @@ def is_valid_indicator_grid_fromversion(self):
                 return False
 
         return True
+
+    def is_valid_incident_field_name_prefix(self):
+        ignored_packs = ['Common Types']
+        pack_metadata = get_pack_metadata(self.file_path)
+        pack_name = pack_metadata.get("name")
+        name_prefixes = pack_metadata.get('itemPrefix', [])
```
Comment (roysagi) [inline demisto_sdk/commands/common/hook_validations/incident_field.py:408]:
```suggestion
        name_prefixes = pack_metadata.get('itemPrefix', []) if pack_metadata.get('itemPrefix') else [pack_name]
```

File: demisto_sdk/commands/common/hook_validations/incident_field.py [hunk 4]
```diff
@@ -397,3 +400,24 @@ def is_valid_indicator_grid_fromversion(self):
...
+        if pack_name and pack_name not in ignored_packs:
+            for prefix in name_prefixes:
+                if self.current_file.get("name", "").startswith(prefix):
+                    return True
+            error_message, error_code = Errors.invalid_incident_field_prefix(field_name)
+            if self.handle_error(
+                    error_message,
+                    error_code,
+                    file_path=self.file_path,
+                    suggested_fix=Errors.suggest_fix_field_name(field_name, pack_name)):
```
Comment (roysagi) [inline demisto_sdk/commands/common/hook_validations/incident_field.py:421]: Well done!

File: demisto_sdk/commands/common/hook_validations/incident_field.py [hunk 5]
```diff
@@ -397,3 +400,24 @@ def is_valid_indicator_grid_fromversion(self):
...
+        pack_metadata = get_pack_metadata(self.file_path)
+        pack_name = pack_metadata.get("name")
+        name_prefixes = pack_metadata.get('itemPrefix', [])
+        if not name_prefixes:
+            name_prefixes = [pack_name]
+        field_name = self.current_file.get("name", "")
+        if pack_name and pack_name not in ignored_packs:
+            for prefix in name_prefixes:
+                if self.current_file.get("name", "").startswith(prefix):
+                    return True
```
Comment (roysagi) [inline demisto_sdk/commands/common/hook_validations/incident_field.py:415]:
```suggestion
                    return True
                    
```

File: demisto_sdk/commands/common/hook_validations/incident_field.py [hunk 6]
```diff
@@ -397,3 +400,24 @@ def is_valid_indicator_grid_fromversion(self):
                 return False
 
         return True
+
+    def is_valid_incident_field_name_prefix(self):
```
Comment (roysagi) [inline demisto_sdk/commands/common/hook_validations/incident_field.py:404]: this function is not trivial - please add a docstring here.

File: demisto_sdk/commands/common/hook_validations/incident_field.py [hunk 7]
```diff
@@ -397,3 +400,24 @@ def is_valid_indicator_grid_fromversion(self):
...
+            for prefix in name_prefixes:
+                if self.current_file.get("name", "").startswith(prefix):
+                    return True
+            error_message, error_code = Errors.invalid_incident_field_prefix(field_name)
+            if self.handle_error(
+                    error_message,
+                    error_code,
+                    file_path=self.file_path,
+                    suggested_fix=Errors.suggest_fix_field_name(field_name, pack_name)):
+                return False
```
Comment (roysagi) [inline demisto_sdk/commands/common/hook_validations/incident_field.py:422]:
```suggestion
                return False
                
```

File: demisto_sdk/commands/common/hook_validations/incident_field.py [hunk 8]
```diff
@@ -397,3 +400,24 @@ def is_valid_indicator_grid_fromversion(self):
...
 
         return True
+
+    def is_valid_incident_field_name_prefix(self):
+        ignored_packs = ['Common Types']
+        pack_metadata = get_pack_metadata(self.file_path)
+        pack_name = pack_metadata.get("name")
+        name_prefixes = pack_metadata.get('itemPrefix', [])
+        if not name_prefixes:
+            name_prefixes = [pack_name]
```
Comment (roysagi) [inline demisto_sdk/commands/common/hook_validations/incident_field.py:410]: you can remove this and see above

File: demisto_sdk/commands/common/tests/incident_field_test.py [hunk 9]
```diff
@@ -352,3 +352,39 @@ def test_is_valid_grid_fromversion(self, field_type, from_version, file_type, is
...
+    INPUTS_NAMES2 = [
+        (FIELD_NAME1, PACK_METADATA1, False),
+        (FIELD_NAME1, PACK_METADATA2, True),
+        (FIELD_NAME2, PACK_METADATA1, True),
+        (FIELD_NAME2, PACK_METADATA2, False),
+        (FIELD_NAME3, PACK_METADATA1, False)
+    ]
+
+    @pytest.mark.parametrize('current_file,pack_metadata, answer', INPUTS_NAMES2)
+    def test_is_valid_name_prefix(self, current_file, pack_metadata, answer, mocker):
```
Comment (roysagi) [inline demisto_sdk/commands/common/tests/incident_field_test.py:377]: please add given-when-then docstring

File: demisto_sdk/commands/validate/validate_manager.py [hunk 10]
```diff
@@ -634,10 +634,10 @@ def validate_incident_field(self, structure_validator, pack_error_ignore_list, i
                                                           print_as_warnings=self.print_ignored_errors,
                                                           json_file_path=self.json_file_path)
         if is_modified and self.is_backward_check:
-            return all([incident_field_validator.is_valid_file(validate_rn=False),
+            return all([incident_field_validator.is_valid_file(validate_rn=False, is_new_file=True, validate_all=self.validate_all),
```
Comment (roysagi) [inline demisto_sdk/commands/validate/validate_manager.py:637]: `is_new_file = not is_modified` no?

File: demisto_sdk/commands/validate/validate_manager.py [hunk 11]
```diff
@@ -634,10 +634,10 @@ def validate_incident_field(self, structure_validator, pack_error_ignore_list, i
...
                         incident_field_validator.is_backward_compatible()])
         else:
-            return incident_field_validator.is_valid_file(validate_rn=False)
+            return incident_field_validator.is_valid_file(validate_rn=False, is_new_file=True, validate_all=self.validate_all)
```
Comment (roysagi) [inline demisto_sdk/commands/validate/validate_manager.py:640]: same comment as above

File: demisto_sdk/commands/common/errors.py [hunk 12]
```diff
@@ -1661,3 +1662,16 @@ def xsoar_config_file_is_not_json(file_path):
...
+    @error_code_decorator
+    def invalid_incident_field_prefix(field_name):
+        return f"Field name: {field_name} is invalid. Field name must start with the relevant pack name."
+
+
+    @staticmethod
+    def suggest_fix_field_name(field_name, pack_name):
+        return f"To fix the problem, add pack name prefix to the field name." \
+               f"you can use the pack name or one of the prefixes in packe.metadat itemPrefix. Example: {pack_name} {field_name}." \
+               "Also make sure to update the field id and cliName accordingly." \
```
Comment (roysagi) [inline demisto_sdk/commands/common/errors.py:1676]: suggest how the cli should look as well please

**Review by roysagi — CHANGES_REQUESTED** [review]

File: demisto_sdk/commands/validate/validate_manager.py (hunk 10 above)
Comment (sabaghadar) [reply demisto_sdk/commands/validate/validate_manager.py:637]: right.

File: demisto_sdk/commands/common/errors.py [hunk 13]
```diff
@@ -16,12 +16,11 @@
     'MP106',
     'PA113', 'PA116', 'PA124', 'PA125',
     'PB104', 'PB105', 'PB106', 'PB110', 'PB111', 'PB112'
-    'RM100', 'RM102', 'RM104', 'RM106',
+                                                 'RM100', 'RM102', 'RM104', 'RM106',
```
Comment (roysagi) [inline demisto_sdk/commands/common/errors.py:19]: revert this change please

**Review by roysagi — APPROVED** [review]
Seems good - please add a changelog entry

**Pull request merged**